### PR TITLE
add quickhighlight plugin

### PR DESCRIPTION
**Pull request by FreaxMATE**
This adds a Quickhighlight plugin highlighting every occurrences of the selected text. Adapted from the correspondent [gedit plugin](https://gitlab.gnome.org/GNOME/gedit/-/tree/master/plugins/quickhighlight)

plugin

### CODE REVIEW

File: plugins/quickhighlight/pluma-quickhighlight-plugin.c [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+        return G_SOURCE_REMOVE;
+    }
+
+    if (plugin->priv->search_context == NULL)
+    {
+        search_settings =
+            g_object_new (GTK_SOURCE_TYPE_SEARCH_SETTINGS,
+                          "at-word-boundaries", FALSE,
+                          "case-sensitive", TRUE,
+                          "regex-enabled", FALSE,
```
Comment (vkareh) [inline plugins/quickhighlight/pluma-quickhighlight-plugin.c:131]: I would love for these to be settings exposed either through the plugin Preferences UI or some gsettings

**Review by vkareh — CHANGES_REQUESTED** [review]
This builds well and works when loading it. However, I get this error when closing Pluma after having the plugin enabled:
```
sys:1: Warning: invalid unclassed pointer in cast to 'PlumaQuickHighlightPlugin'
**
ERROR:pluma-quickhighlight-plugin.c:459:pluma_quick_highlight_plugin_notify_buffer_cb: assertion failed: (PLUMA_IS_QUICK_HIGHLIGHT_PLUGIN (plugin))
Bail out! ERROR:pluma-quickhighlight-plugin.c:459:pluma_quick_highlight_plugin_notify_buffer_cb: assertion failed: (PLUMA_IS_QUICK_HIGHLIGHT_PLUGIN (plugin))
Aborted (core dumped)
```

**Review by mbkma — APPROVED** [review]
With the newest commit https://github.com/mate-desktop/pluma/commit/c7d460d479891b305c96b3de18817c541929d0a8 to pluma, this works fine. Thanks for contributing!

File: plugins/quickhighlight/pluma-quickhighlight-plugin.c (hunk 1 above)
Comment (mbkma) [reply plugins/quickhighlight/pluma-quickhighlight-plugin.c:131]: I personally do not need this to be configurable. But if someone likes to do it, a PR would be nice of course.

**Comment by mbkma** [comment]
@vkareh Do you still get the error with the most recent pluma commit?

**Review by rbuj — COMMENTED** [review]
You have to modify po/POTFILES.in to include the files that have localizable strings (l10n).

**Comment by FreaxMATE** [comment]
> You have to modify po/POTFILES.in to include the files that have localizable strings (l10n).

Fixed.

File: configure.ac [hunk 2]
```diff
@@ -86,7 +86,7 @@ GLIB_GSETTINGS
 # ================================================================
 
 # C plugins that don't need special dependencies
-C_PLUGINS="bookmarks sourcecodebrowser wordcompletion"
+C_PLUGINS="bookmarks sourcecodebrowser wordcompletion quickhighlight"
```
Comment (rbuj) [inline configure.ac]: I think you can keep the plugins sorted by name

File: configure.ac [hunk 3]
```diff
@@ -229,6 +229,9 @@ plugins/sourcecodebrowser/sourcecodebrowser.plugin.desktop.in
 plugins/wordcompletion/org.mate.pluma.plugins.wordcompletion.gschema.xml
 plugins/wordcompletion/wordcompletion.plugin.desktop.in
 plugins/wordcompletion/Makefile
+plugins/quickhighlight/quickhighlight.plugin.desktop.in
+plugins/quickhighlight/Makefile
+plugins/quickhighlight/pluma-quickhighlight.metainfo.xml.in
```
Comment (rbuj) [inline configure.ac]: I think you can keep the files sorted by name

File: plugins/quickhighlight/pluma-quickhighlight-plugin.c [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+pluma_quick_highlight_plugin_unref_weak_buffer (PlumaQuickHighlightPlugin *plugin)
+{
+    g_return_if_fail (PLUMA_IS_QUICK_HIGHLIGHT_PLUGIN (plugin));
+
+    if (plugin->priv->buffer == NULL)
+    {
+        return;
+    }
+
+    if (plugin->priv->delete_range_handler_id > 0)
```
Comment (rbuj) [inline plugins/quickhighlight/pluma-quickhighlight-plugin.c]: `plugin->priv->delete_range_handler_id` is an unsigned long integer, so you must check that the value is different from zero, check the other identifiers (_handler_id).

File: plugins/quickhighlight/pluma-quickhighlight-plugin.c [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+
+    if (plugin->priv->buffer != NULL)
+    {
+        g_object_weak_ref (G_OBJECT (plugin->priv->buffer),
+                           pluma_quick_highlight_plugin_notify_weak_buffer_cb,
+                           plugin);
+
+        plugin->priv->style_scheme_handler_id =
+            g_signal_connect (plugin->priv->buffer,
+                              "notify::style-scheme",
```
Comment (rbuj) [inline plugins/quickhighlight/pluma-quickhighlight-plugin.c]: we usually put the signal name on the same line for the signal source in g_signal_connect.

File: plugins/quickhighlight/pluma-quickhighlight-plugin.c (hunk 4 above)
Comment (FreaxMATE) [reply plugins/quickhighlight/pluma-quickhighlight-plugin.c]: It checks whether the value is greater than zero or what do you mean?

Comment (rbuj) [reply plugins/quickhighlight/pluma-quickhighlight-plugin.c]: An unsigned long cannot be lower than zero.
```C
if (handler_id != 0)
  {
    g_signal_handler_disconnect (instance, handler_id);
    handler_id = 0;
  }
```
or
```C
if (handler_id)
  {
    g_signal_handler_disconnect (instance, handler_id);
    handler_id = 0;
  }
```
or
```C
#if GLIB_CHECK_VERSION(2,62,0)
g_clear_signal_handler (&handler_id, instance);
#else
if (handler_id != 0)
  {
    g_signal_handler_disconnect (instance, handler_id);
    handler_id = 0;
  }
#endif
```
or
```C
#if GLIB_CHECK_VERSION(2,62,0)
g_clear_signal_handler (&handler_id, instance);
#else
if (handler_id)
  {
    g_signal_handler_disconnect (instance, handler_id);
    handler_id = 0;
  }
#endif
```

**Review by rbuj — APPROVED** [review]
LGTM

**Review by raveit65 — DISMISSED** [review]
I get an core dump with fedora abrt-tool when i close pluma and the plugin is enabled.
```
[rave@mother ~]$ pluma
sys:1: Warning: invalid unclassed pointer in cast to 'PlumaQuickHighlightPlugin'
**
ERROR:pluma-quickhighlight-plugin.c:456:pluma_quick_highlight_plugin_notify_buffer_cb: assertion failed: (PLUMA_IS_QUICK_HIGHLIGHT_PLUGIN (plugin))
Bail out! ERROR:pluma-quickhighlight-plugin.c:456:pluma_quick_highlight_plugin_notify_buffer_cb: assertion failed: (PLUMA_IS_QUICK_HIGHLIGHT_PLUGIN (plugin))
Aborted (core dumped)
[rave@mother ~]$ 
```

**Comment by FreaxMATE** [comment]
> I get an core dump with fedora abrt-tool when i close pluma and the plugin is enabled.
> 
> ```
> [rave@mother ~]$ pluma
> sys:1: Warning: invalid unclassed pointer in cast to 'PlumaQuickHighlightPlugin'
> **
> ERROR:pluma-quickhighlight-plugin.c:456:pluma_quick_highlight_plugin_notify_buffer_cb: assertion failed: (PLUMA_IS_QUICK_HIGHLIGHT_PLUGIN (plugin))
> Bail out! ERROR:pluma-quickhighlight-plugin.c:456:pluma_quick_highlight_plugin_notify_buffer_cb: assertion failed: (PLUMA_IS_QUICK_HIGHLIGHT_PLUGIN (plugin))
> Aborted (core dumped)
> [rave@mother ~]$ 
> ```

Did you compile pluma from master as well? https://github.com/mate-desktop/pluma/pull/651

**Comment by raveit65** [comment]
I didn't rebuild pluma 1.26.0.

**Comment by raveit65** [comment]
After updating pluma the issue with closing pluma is gone.
But we have a new build warning.
https://app.travis-ci.com/github/mate-desktop/pluma-plugins/jobs/553730879#L676
```
plugins/quickhighlight/pluma-quickhighlight-plugin.c:284:5: error: syntax error [syntaxError]

    if G_LIKELY (mark != plugin->priv->insert_mark)

    ^
```
Any chance to fix this before merging?

**Comment by FreaxMATE** [comment]
> After updating pluma the issue with closing pluma is gone. But we have a new build warning. https://app.travis-ci.com/github/mate-desktop/pluma-plugins/jobs/553730879#L676
> 
> ```
> plugins/quickhighlight/pluma-quickhighlight-plugin.c:284:5: error: syntax error [syntaxError]
> 
>     if G_LIKELY (mark != plugin->priv->insert_mark)
> 
>     ^
> ```
> 
> Any chance to fix this before merging?

fixed

**Comment by raveit65** [comment]
New warning.
https://app.travis-ci.com/github/mate-desktop/pluma-plugins/jobs/554953082#L675
```
plugins/quickhighlight/pluma-quickhighlight-plugin.c:277:58: style: Parameter 'mark' can be declared with const [constParameter]

                                          GtkTextMark   *mark,

                                                         ^
```
Any chance to look for yourself in travis CI build logs?

**Comment by raveit65** [comment]
@mbkma 
I don't really mind about this new warning. Feel free to merge it.

**Comment by mbkma** [comment]
LGTM, the warning is not a real issue.

**Pull request merged**